File: 15_Rambunctious_Recitation/game.py
#!/usr/bin/env python3

import sys

MY_INPUT = "14,1,17,0,3,20"
# Part one
# TURN_IN_QUESTION = 2020

# Part two
# Terminates after ~30 seconds
TURN_IN_QUESTION = 30000000

data = sys.argv[1] if len(sys.argv) > 1 else MY_INPUT
starting_numbers = [int(x) for x in data.split(",")]

last_spoken = dict()
last_number = None

for turn in range(1, TURN_IN_QUESTION + 1):
    if turn - 1 < len(starting_numbers):
        number_spoken = starting_numbers[turn - 1]
    else:
        if last_number in last_spoken and last_spoken[last_number][0] is not None:
            number_spoken = (turn - 1) - last_spoken[last_number][0]
        else:
            number_spoken = 0

    if number_spoken in last_spoken:
        last_spoken[number_spoken] = (last_spoken[number_spoken][1], turn)
    else:
        last_spoken[number_spoken] = (None, turn)
    last_number = number_spoken

print(last_number)

File: 15_Rambunctious_Recitation/game.rs
use std::collections::HashMap;
use std::env;
use std::num::NonZeroUsize;
use std::str::FromStr;

const MY_INPUT: &'static str = "14,1,17,0,3,20";
const TURN_IN_QUESTION: usize = 30000000;
// const TURN_IN_QUESTION: usize = 2020;

fn main() {
    let starting_numbers = env::args()
        .skip(1)
        .next()
        .unwrap_or_else(|| MY_INPUT.to_owned())
        .split(',')
        .map(usize::from_str)
        .collect::<Result<Vec<_>, _>>()
        .expect("Couldn't parse input");
    let mut last_number = 0;
    let mut last_spoken = HashMap::<usize, (Option<NonZeroUsize>, NonZeroUsize)>::new();

    for turn in 1..=TURN_IN_QUESTION {
        let number_spoken: usize = starting_numbers.get(turn - 1).copied().unwrap_or_else(|| {
            if let Some((Some(before_last), last)) = last_spoken.get(&last_number) {
                last.get() - before_last.get()
            } else {
                0
            }
        });

        last_spoken
            .entry(number_spoken)
            .and_modify(|entry| {
                *entry = (Some(entry.1), NonZeroUsize::new(turn).unwrap());
            })
            .or_insert((None, NonZeroUsize::new(turn).unwrap()));

        last_number = number_spoken;
    }

    println!("{}", last_number);
}
